File: src/bot/chat_manager.py
"""ChatManager：管理 Chat → Agent 映射"""
from typing import Dict, Optional, Callable, Awaitable
import asyncio
import time
import logging
from src.claude.agent import Agent

logger = logging.getLogger(__name__)


class ChatManager:
    """管理多个 Chat 的 Agent 实例"""

    def __init__(self):
        from src.config import config
        self.chats: Dict[str, dict] = {}  # chat_id -> {cwd, session_id, session_notified}
        self.agents: Dict[str, Agent] = {}  # chat_id -> Agent
        self.current_reactions: Dict[str, tuple] = {}  # chat_id -> (message_id, reaction_id)
        self.default_cwd = config.claude.work_root
        self.start_time = time.time()
        self.response_complete_callback: Optional[Callable[[], None]] = None

    async def start(self):
        """启动 ChatManager"""
        logger.info('ChatManager started')

    async def stop(self):
        """停止 ChatManager，清理所有 Agent"""
        for chat_id, agent in list(self.agents.items()):
            try:
                # 增加超时时间，给 Windows 管道更多关闭时间
                await asyncio.wait_for(agent.destroy(), timeout=5.0)
            except asyncio.TimeoutError:
                logger.warning('Agent destroy timeout', extra={'chat_id': chat_id})
            except Exception as e:
                logger.error('Error destroying agent', extra={
                    'chat_id': chat_id,
                    'error': str(e)
                })

        self.agents.clear()
        logger.info('ChatManager stopped')

    def on_response_complete(self, callback: Callable[[], None]):
        """注册响应完成回调"""
        self.response_complete_callback = callback

    def get_or_create_agent(self, chat_id: str) -> tuple[Agent, str | None]:
        """
        获取或创建 Agent

        Returns:
            (agent, expected_session_id): Agent 实例和期望的 session ID
        """
        agent = self.agents.get(chat_id)

        # 如果已存在，直接返回
        if agent:
            return agent, agent.get_session_id()

        # 创建新 Agent
        chat_data = self.chats.get(chat_id, {})
        cwd = chat_data.get('cwd', self.default_cwd)
        session_id = chat_data.get('session_id')

        # 定义消息开始处理回调
        async def on_message_start(message_id: str):
            # 移除旧 reaction
            if chat_id in self.current_reactions:
                old_message_id, old_reaction_id = self.current_reactions[chat_id]
                from src.services.message_service import message_service
                try:
                    message_service.remove_reaction(old_message_id, old_reaction_id)
                except Exception as e:
                    logger.warning('Failed to remove old reaction', extra={
                        'chat_id': chat_id,
                        'error': str(e)
                    })

            # 添加新 reaction
            from src.services.message_service import message_service
            try:
                reaction_id = message_service.add_reaction(message_id, 'Typing')
                self.current_reactions[chat_id] = (message_id, reaction_id)
                logger.debug('Added reaction', extra={
                    'chat_id': chat_id,
                    'message_id': message_id
                })
            except Exception as e:
                logger.warning('Failed to add reaction', extra={
                    'chat_id': chat_id,
                    'error': str(e)
                })

        # 定义响应回调
        async def on_response(response_text: str):
            # 移除 reaction
            if chat_id in self.current_reactions:
                message_id, reaction_id = self.current_reactions.pop(chat_id)
                from src.services.message_service import message_service
                try:
                    message_service.remove_reaction(message_id, reaction_id)
                except Exception as e:
                    logger.warning('Failed to remove reaction', extra={
                        'chat_id': chat_id,
                        'error': str(e)
                    })

            # 检查 session 是否变化（仅首次响应）
            chat_data = self.chats.get(chat_id, {})
            if not chat_data.get('session_notified', False):
                actual_session_id = agent.get_session_id()
                session_changed = actual_session_id != session_id

                # 生成通知消息
                notification = None
                if not session_id:
                    # 首次创建会话
                    notification = f'🆕 新会话: {agent.get_cwd()} ({actual_session_id})'
                elif session_changed:
                    # 恢复失败
                    notification = f'⚠️ 恢复失败，已创建新会话: {agent.get_cwd()} ({actual_session_id})'

                # 发送通知
                if notification:
                    from src.services.message_service import message_service
                    try:
                        await message_service.send_text_message(chat_id, notification)
                        logger.info('Session change notification sent', extra={
                            'chat_id': chat_id,
                            'session_changed': session_changed
                        })
                    except Exception as e:
                        logger.error('Failed to send session notification', extra={
                            'chat_id': chat_id,
                            'error': str(e)
                        })

                # 标记已通知
                self.chats[chat_id]['session_notified'] = True

            # 发送响应消息
            from src.services.message_service import message_service
            try:
                await message_service.send_text_message(chat_id, response_text)
            except Exception as e:
                logger.error('Failed to send response', extra={
                    'chat_id': chat_id,
                    'error': str(e)
                })

        try:
            agent = Agent(chat_id, cwd, session_id, on_message_start, on_response)
            self.agents[chat_id] = agent

            # 初始化 chat 数据（重置 session_notified）
            if chat_id not in self.chats:
                self.chats[chat_id] = {'cwd': cwd, 'session_id': session_id, 'session_notified': False}
            else:
                self.chats[chat_id]['session_notified'] = False

            # 返回期望的 session_id（用于后续对比）
            return agent, session_id
        except Exception as e:
            logger.error('Failed to create agent', extra={
                'chat_id': chat_id,
                'error': str(e)
            })
            raise RuntimeError(f'创建 Agent 失败: {e}')

    async def send_message(
        self,
        chat_id: str,
        message_id: str,
        text: str
    ):
        """
        发送消息

        Args:
            chat_id: 会话 ID
            message_id: 消息 ID
            text: 消息内容
        """
        agent, _ = self.get_or_create_agent(chat_id)

        try:
            # 调用 agent（消息入队，立即返回）
            await agent.send_message(message_id, text)

            # 更新 chat 数据中的 session_id
            if agent.session_id:
                if chat_id not in self.chats:
                    self.chats[chat_id] = {}
                self.chats[chat_id]['session_id'] = agent.session_id
                self.chats[chat_id]['cwd'] = agent.cwd

            # 触发回调
            if self.response_complete_callback:
                self.response_complete_callback()

        except asyncio.CancelledError:
            logger.info('Message processing cancelled', extra={'chat_id': chat_id})
            raise

        except Exception as e:
            logger.error('Error in send_message', extra={
                'chat_id': chat_id,
                'error': str(e)
            })
            raise

    async def interrupt(self, chat_id: str) -> str:
        """尝试中断当前任务"""
        agent = self.agents.get(chat_id)
        if not agent:
            return 'no_session'

        try:
            await asyncio.wait_for(agent.interrupt(), timeout=3.0)

            # 清理 reaction
            if chat_id in self.current_reactions:
                message_id, reaction_id = self.current_reactions.pop(chat_id)
                from src.services.message_service import message_service
                try:
                    message_service.remove_reaction(message_id, reaction_id)
                except Exception as e:
                    logger.warning('Failed to remove reaction after interrupt', extra={
                        'chat_id': chat_id,
                        'error': str(e)
                    })

            return 'success'
        except asyncio.TimeoutError:
            logger.warning('Interrupt timeout', extra={'chat_id': chat_id})
            return 'timeout'
        except Exception as e:
            logger.error('Interrupt failed', extra={
                'chat_id': chat_id,
                'error': str(e)
            })
            return 'error'

    async def reset(self, chat_id: str) -> str:
        """重置会话（强制清理）"""
        agent = self.agents.get(chat_id)

        if agent:
            try:
                await asyncio.wait_for(agent.destroy(), timeout=2.0)
            except asyncio.TimeoutError:
                logger.warning('Agent destroy timeout', extra={'chat_id': chat_id})
            except Exception as e:
                logger.error('Error destroying agent', extra={
                    'chat_id': chat_id,
                    'error': str(e)
                })

            # 强制清理
            self.agents.pop(chat_id, None)

        # 保留 cwd，清空 session_id 和通知标志
        cwd = self.chats.get(chat_id, {}).get('cwd', self.default_cwd)
        self.chats[chat_id] = {'cwd': cwd, 'session_id': None, 'session_notified': False}

        logger.info('Session reset', extra={'chat_id': chat_id})
        return cwd

    async def switch_cwd(self, chat_id: str, new_cwd: str):
        """切换工作目录（强制清理 Agent）"""
        current_cwd = self.chats.get(chat_id, {}).get('cwd', self.default_cwd)
        if current_cwd == new_cwd:
            logger.debug('Cwd unchanged', extra={
                'chat_id': chat_id,
                'cwd': new_cwd
            })
            return

        agent = self.agents.get(chat_id)
        if agent:
            try:
                await asyncio.wait_for(agent.destroy(), timeout=2.0)
            except asyncio.TimeoutError:
                logger.warning('Agent destroy timeout when switching cwd', extra={
                    'chat_id': chat_id
                })
            except Exception as e:
                logger.error('Error destroying agent when switching cwd', extra={
                    'chat_id': chat_id,
                    'error': str(e)
                })

            # 强制清理
            self.agents.pop(chat_id, None)

        # 更新 cwd，清空 session_id 和通知标志
        self.chats[chat_id] = {'cwd': new_cwd, 'session_id': None, 'session_notified': False}
        logger.info('Switched cwd', extra={
            'chat_id': chat_id,
            'new_cwd': new_cwd
        })

    def get_session_info(self, chat_id: str) -> str:
        """获取会话信息"""
        agent = self.agents.get(chat_id)
        chat_data = self.chats.get(chat_id, {})

        if not agent:
            cwd = chat_data.get('cwd', self.default_cwd)
            return f'当前没有活跃的会话\n工作目录: {cwd}'

        session_id = agent.get_session_id() or '无'
        cwd = agent.get_cwd()
        uptime = time.time() - agent.get_start_time()

        return (
            f'会话信息:\n'
            f'- Session ID: {session_id[:16]}...\n'
            f'- 工作目录: {cwd}\n'
            f'- 运行时长: {int(uptime)}秒'
        )

    def get_debug_info(self) -> str:
        """获取调试信息"""
        uptime = time.time() - self.start_time
        info = [
            '**系统状态**',
            f'- 运行时长: {int(uptime)}秒',
            f'- 活跃会话: {len(self.chats)}',
            f'- 活跃 Agent: {len(self.agents)}',
            '',
            '**会话列表**',
        ]

        for chat_id, data in self.chats.items():
            agent = self.agents.get(chat_id)
            session_id = data.get('session_id', '无')
            if session_id and session_id != '无':
                session_id = session_id[:8] + '...'

            if agent:
                agent_id = agent.get_agent_id()
                # 截断 agent_id，只保留前 12 个字符
                agent_id_short = agent_id[:12] + '...' if len(agent_id) > 12 else agent_id

                # 获取状态（优先显示更重要的状态）
                if not agent._connected:
                    status = '未连接'
                elif agent.is_busy():
                    status = '处理中'
                else:
                    status = '空闲'

                info.append(
                    f'- Chat: `{chat_id[:8]}...`\n'
                    f'  - Agent: {status} `{agent_id_short}`\n'
                    f'  - Session: `{session_id}`\n'
                    f'  - CWD: `{data.get("cwd", "N/A")}`'
                )
            else:
                info.append(
                    f'- Chat: `{chat_id[:8]}...`\n'
                    f'  - Agent: 未启动\n'
                    f'  - Session: `{session_id}`\n'
                    f'  - CWD: `{data.get("cwd", "N/A")}`'
                )

        return '\n'.join(info)


# 单例
chat_manager = ChatManager()

File: apps/typescript/src/bot/chat-manager.ts
import logger from '../utils/logger';
import messageService from '../services/message.service';
import { createAgent, type CreateAgentOptions } from '../agent/factory';
import type { AgentProvider } from '../config';
import type { ChatAgent, SendMessageOptions } from '../agent/types';
import type { DirectorySummary, SessionSummary, SessionTarget } from '../agent/session-history';
import {
  getRecentDirectories as getClaudeRecentDirectories,
  getSessionList as getClaudeSessionList,
  getValidSessions as getClaudeValidSessions,
} from '../claude/session-scanner';
import {
  getRecentDirectories as getCodexRecentDirectories,
  getSessionList as getCodexSessionList,
  getValidSessions as getCodexValidSessions,
} from '../codex/session-scanner';
import config from '../config';
import { chatBindingStore, type ChatBindingStore } from './chat-binding-store';

interface ChatData {
  cwd: string;
  provider: AgentProvider;
  sessionId: string | undefined;
  sessionNotified?: boolean;
}

type AgentFactory = (options: CreateAgentOptions) => ChatAgent;

interface ChatManagerOptions {
  bindingStore?: ChatBindingStore;
  defaultCwd?: string;
  defaultProvider?: AgentProvider;
  agentFactory?: AgentFactory;
}

function formatDuration(seconds: number): string {
  const wholeSeconds = Math.max(0, Math.floor(seconds));
  if (wholeSeconds < 60) {
    return `${wholeSeconds}秒`;
  }
  if (wholeSeconds < 3600) {
    const minutes = Math.floor(wholeSeconds / 60);
    const secs = wholeSeconds % 60;
    return `${minutes}分${secs}秒`;
  }
  if (wholeSeconds < 86400) {
    const hours = Math.floor(wholeSeconds / 3600);
    const minutes = Math.floor((wholeSeconds % 3600) / 60);
    return `${hours}小时${minutes}分`;
  }
  const days = Math.floor(wholeSeconds / 86400);
  const hours = Math.floor((wholeSeconds % 86400) / 3600);
  return `${days}天${hours}小时`;
}

export class ChatManager {
  private chats = new Map<string, ChatData>();
  private agents = new Map<string, ChatAgent>();
  private readonly bindingStore: ChatBindingStore;
  private readonly defaultCwd: string;
  private readonly defaultProvider: AgentProvider;
  private readonly agentFactory: AgentFactory;
  private readonly startTime: number;

  constructor(options: ChatManagerOptions = {}) {
    this.bindingStore = options.bindingStore ?? chatBindingStore;
    this.defaultCwd = options.defaultCwd ?? config.claude.workRoot;
    this.defaultProvider = options.defaultProvider ?? config.agent.provider;
    this.agentFactory = options.agentFactory ?? createAgent;
    this.startTime = Date.now();
  }

  async start(): Promise<void> {
    logger.info('[ChatManager] Started', { provider: this.defaultProvider });
  }

  getProvider(chatId?: string): AgentProvider {
    if (!chatId) {
      return this.defaultProvider;
    }
    return this.getChatProvider(chatId);
  }

  getCurrentCwd(chatId: string): string {
    return this.getChatCwd(chatId);
  }

  supportsSessionResume(chatId: string): boolean {
    const provider = this.getChatProvider(chatId);
    return provider === 'claude' || provider === 'codex';
  }

  async sendMessage(chatId: string, text: string, options?: SendMessageOptions): Promise<void> {
    const provider = this.getChatProvider(chatId);
    const agent = this.getOrCreateAgent(chatId);
    logger.info('[ChatManager] Sending message', {
      chatId,
      provider,
      agentId: agent.getAgentId(),
      sessionId: agent.getSessionId(),
      messageLength: text.length,
      messageText: text,
    });
    await agent.sendMessage(text, options);
  }

  async interrupt(chatId: string): Promise<'success' | 'timeout' | 'no_session' | 'error'> {
    const agent = this.agents.get(chatId);
    if (!agent) {
      logger.warn('[ChatManager] No agent to interrupt', { chatId });
      return 'no_session';
    }

    try {
      const sent = agent.interrupt();
      const result = sent ? 'success' : 'error';
      logger.info('[ChatManager] Interrupt result', {
        chatId,
        provider: this.getChatProvider(chatId),
        agentId: agent.getAgentId(),
        result,
      });
      return result;
    } catch (error: any) {
      logger.error('[ChatManager] Interrupt failed', { chatId, error: error.message });
      return 'error';
    }
  }

  async switchCwd(chatId: string, newCwd: string): Promise<void> {
    const data = this.chats.get(chatId);
    const currentBindingCwd = this.getStoredCwd(chatId);
    const currentCwd = data?.cwd ?? currentBindingCwd ?? this.defaultCwd;
    const provider = data?.provider ?? this.defaultProvider;
    if (currentCwd === newCwd && currentBindingCwd === newCwd) {
      logger.debug('[ChatManager] Cwd unchanged, skipping switch', { chatId, cwd: newCwd });
      return;
    }

    await this.destroyAgent(chatId, 'switching cwd');
    this.bindingStore.set(chatId, newCwd);
    this.chats.set(chatId, { cwd: newCwd, provider, sessionId: undefined, sessionNotified: false });
    logger.info('[ChatManager] Switched cwd', { chatId, oldCwd: currentCwd, newCwd, provider });
  }

  async switchProvider(chatId: string, provider: AgentProvider): Promise<{ changed: boolean; cwd: string }> {
    const data = this.chats.get(chatId);
    const currentProvider = data?.provider ?? this.defaultProvider;
    const cwd = data?.cwd ?? this.getStoredCwd(chatId) ?? this.defaultCwd;

    if (currentProvider === provider) {
      logger.debug('[ChatManager] Provider unchanged, skipping switch', { chatId, provider });
      return { changed: false, cwd };
    }

    await this.destroyAgent(chatId, 'switching provider');
    this.chats.set(chatId, {
      cwd,
      provider,
      sessionId: undefined,
      sessionNotified: false,
    });
    logger.info('[ChatManager] Switched provider', {
      chatId,
      oldProvider: currentProvider,
      newProvider: provider,
      cwd,
    });
    return { changed: true, cwd };
  }

  async reset(chatId: string): Promise<string> {
    await this.destroyAgent(chatId, 'reset');

    const cwd = this.getChatCwd(chatId);
    const provider = this.getChatProvider(chatId);
    this.chats.set(chatId, { cwd, provider, sessionId: undefined, sessionNotified: false });
    logger.info('[ChatManager] Session reset', { chatId, cwd, provider });
    return cwd;
  }

  getSessionInfo(chatId: string): string {
    const agent = this.agents.get(chatId);
    const data = this.chats.get(chatId);
    const provider = data?.provider ?? this.defaultProvider;

    if (!agent) {
      const cwd = data?.cwd ?? this.getStoredCwd(chatId) ?? this.defaultCwd;
      return `当前没有活跃会话\nProvider: ${provider}\n工作目录: ${cwd}`;
    }

    const sessionId = agent.getSessionId() || '无';
    const cwd = agent.getCwd();
    const uptime = (Date.now() - agent.getStartTime()) / 1000;

    return [
      '会话信息:',
      `- Provider: ${provider}`,
      `- Session ID: ${sessionId === '无' ? sessionId : `${sessionId.slice(0, 16)}...`}`,
      `- 工作目录: ${cwd}`,
      `- 运行时长: ${formatDuration(uptime)}`,
    ].join('\n');
  }

  listSessions(chatId: string): string {
    const provider = this.getChatProvider(chatId);
    if (!this.supportsSessionResume(chatId)) {
      const cwd = this.getChatCwd(chatId);
      return [
        `当前 provider: ${provider}`,
        `工作目录: \`${cwd}\``,
        '',
        '当前 provider 暂不支持 /resume 和历史会话列表。',
      ].join('\n');
    }

    const chatData = this.chats.get(chatId);
    const cwd = this.getChatCwd(chatId);
    const sessions = this.getValidSessionsForChat(chatId);
    const totalSessions = this.getSessionCount(chatId);

    if (sessions.length === 0) {
      return `工作目录: \`${cwd}\`\n\n暂无 session 记录`;
    }

    const currentSession = chatData?.sessionId;
    const lines: string[] = [];

    lines.push(`**Sessions（最近 ${sessions.length} 条）**`);
    lines.push(`工作目录: \`${cwd}\``);
    if (cwd === this.defaultCwd) {
      lines.push('当前为默认目录，已显示所有目录的最近会话。');
    }
    if (totalSessions > sessions.length) {
      lines.push(`仅显示最近 ${sessions.length} 条；更多会话请使用 /resume <session_id>。`);
    }
    lines.push('');

    sessions.forEach((session, index) => {
      lines.push(this.formatSessionLine(index + 1, session, currentSession));
      if (session.cwd !== cwd) {
        lines.push(`目录 ${session.cwd}`);
      }
      lines.push(`摘要 ${session.firstMessage}`);
      lines.push('');
    });
    return lines.join('\n');
  }

  getRecentSessions(chatId: string, limit = 9): SessionSummary[] {
    if (!this.supportsSessionResume(chatId)) {
      return [];
    }
    return this.getValidSessionsForChat(chatId, limit);
  }

  getSessionCount(chatId: string): number {
    if (!this.supportsSessionResume(chatId)) {
      return 0;
    }
    return this.getSessionListForChat(chatId).length;
  }

  getRecentDirectories(chatId: string, limit = 9): DirectorySummary[] {
    switch (this.getChatProvider(chatId)) {
      case 'codex':
        return getCodexRecentDirectories(limit);
      case 'claude':
      default:
        return getClaudeRecentDirectories(limit);
    }
  }

  resolveResumeTarget(chatId: string, index: number): SessionTarget | null {
    if (!this.supportsSessionResume(chatId)) {
      return null;
    }

    const sessions = this.getSessionListForChat(chatId);
    if (index < 1 || index > sessions.length) {
      return null;
    }

    const target = sessions[index - 1];
    return {
      sessionId: target.sessionId,
      cwd: target.cwd,
    };
  }

  resolveResumeTargetBySessionId(chatId: string, sessionId: string): SessionTarget | null {
    const normalizedId = sessionId.trim();
    if (!normalizedId) {
      return null;
    }

    return this.getSessionListForChat(chatId).find(item => item.sessionId === normalizedId) ?? null;
  }

  async resumeSession(chatId: string, sessionId: string): Promise<string> {
    const provider = this.getChatProvider(chatId);
    if (!this.supportsSessionResume(chatId)) {
      return `当前 provider (${provider}) 暂不支持 /resume。`;
    }

    const target = this.resolveResumeTargetBySessionId(chatId, sessionId);
    if (!target) {
      return `会话不存在: ${sessionId}\n使用 /resume 查看可用的 sessions`;
    }

    await this.destroyAgent(chatId, 'resuming session');
    this.chats.set(chatId, {
      cwd: target.cwd,
      provider,
      sessionId: target.sessionId,
      sessionNotified: false,
    });

    logger.info('[ChatManager] Resumed session', {
      chatId,
      cwd: target.cwd,
      provider,
      sessionId: target.sessionId,
    });
    return `已切换到 session: ${target.sessionId}\n工作目录: ${target.cwd}`;
  }

  getDebugInfo(): string {
    const uptime = (Date.now() - this.startTime) / 1000;
    const info = [
      '**系统状态**',
      `- 默认 Provider: ${this.defaultProvider}`,
      `- 运行时长: ${formatDuration(uptime)}`,
      `- 活跃会话: ${this.chats.size}`,
      `- 活跃 Agent: ${this.agents.size}`,
      '',
      '**会话列表**',
    ];

    if (this.chats.size === 0) {
      info.push('当前没有活跃会话');
    }

    for (const [chatId, data] of this.chats) {
      const agent = this.agents.get(chatId);
      const provider = data.provider;
      let sessionId = data.sessionId || '无';
      if (sessionId !== '无') {
        sessionId = `${sessionId.slice(0, 8)}...`;
      }

      if (agent) {
        const agentId = agent.getAgentId();
        const agentIdShort = agentId.length > 12 ? `${agentId.slice(0, 12)}...` : agentId;
        const status = agent.isInitialized() ? '空闲' : '未连接';

        info.push(
          `- Chat: \`${chatId.slice(0, 8)}...\`\n`
          + `  - Agent: ${status} \`${agentIdShort}\`\n`
          + `  - Provider: ${provider}\n`
          + `  - Session: \`${sessionId}\`\n`
          + `  - CWD: \`${data.cwd}\``,
        );
      } else {
        info.push(
          `- Chat: \`${chatId.slice(0, 8)}...\`\n`
          + '  - Agent: 未启动\n'
          + `  - Provider: ${provider}\n`
          + `  - Session: \`${sessionId}\`\n`
          + `  - CWD: \`${data.cwd}\``,
        );
      }
    }

    return info.join('\n');
  }

  async stop(): Promise<void> {
    for (const agent of this.agents.values()) {
      await agent.destroy();
    }
    this.agents.clear();
    logger.info('[ChatManager] Stopped');
  }

  private getOrCreateAgent(chatId: string): ChatAgent {
    let agent = this.agents.get(chatId);
    const provider = this.getChatProvider(chatId);
    if (agent?.isAlive()) {
      logger.debug('[ChatManager] Reusing existing agent', {
        chatId,
        provider,
        agentId: agent.getAgentId(),
        sessionId: agent.getSessionId(),
      });
      return agent;
    }

    if (agent) {
      logger.info('[ChatManager] Cleaning up dead agent', {
        chatId,
        provider,
        agentId: agent.getAgentId(),
        sessionId: agent.getSessionId(),
      });
      agent.destroy().catch(() => {});
      this.agents.delete(chatId);
    }

    const data = this.chats.get(chatId);
    const cwd = data?.cwd ?? this.getStoredCwd(chatId) ?? this.defaultCwd;
    const storedSessionId = data?.sessionId;
    const resumeSessionId = this.supportsSessionResume(chatId) && storedSessionId
      ? storedSessionId
      : undefined;

    logger.info('[ChatManager] Creating new agent', {
      chatId,
      provider,
      cwd,
      resumeSessionId,
      willResume: !!resumeSessionId,
    });

    agent = this.agentFactory({
      provider,
      chatId,
      cwd,
      resumeSessionId,
    });
    this.agents.set(chatId, agent);

    const expectedSessionId = resumeSessionId ?? agent.getSessionId();

    agent.onResponse((text) => {
      const currentData = this.chats.get(chatId);

      if (currentData && !currentData.sessionNotified) {
        const actualSessionId = agent.getSessionId();
        const sessionChanged = actualSessionId !== expectedSessionId;

        if (sessionChanged || !resumeSessionId) {
          let message = '';
          if (!resumeSessionId) {
            message = `新会话已创建: ${cwd} (${actualSessionId ?? 'pending'})`;
          } else if (sessionChanged) {
            message = `恢复失败，已创建新会话: ${cwd} (${actualSessionId ?? 'pending'})`;
          }

          if (message) {
            messageService.sendTextMessage(chatId, message).catch(() => {});
          }
        }

        this.chats.set(chatId, {
          ...currentData,
          sessionNotified: true,
          sessionId: actualSessionId,
        });
      }

      logger.info('[ChatManager] Agent response received', {
        chatId,
        provider,
        agentId: agent.getAgentId(),
        textLength: text.length,
        responseText: text,
      });
      this.sendResponse(chatId, text);
    });

    agent.onError((error) => {
      logger.error('[ChatManager] Agent error', { chatId, provider, error: error.message });
      messageService.sendTextMessage(chatId, `错误: ${error.message}`).catch(() => {});
    });

    this.chats.set(chatId, { cwd, provider, sessionId: agent.getSessionId(), sessionNotified: false });
    return agent;
  }

  private getChatCwd(chatId: string): string {
    return this.chats.get(chatId)?.cwd ?? this.getStoredCwd(chatId) ?? this.defaultCwd;
  }

  private getChatProvider(chatId: string): AgentProvider {
    return this.chats.get(chatId)?.provider ?? this.defaultProvider;
  }

  private getStoredCwd(chatId: string): string | null {
    return this.bindingStore.get(chatId)?.cwd ?? null;
  }

  private getValidSessionsForChat(chatId: string, limit = 9): SessionSummary[] {
    const cwd = this.getChatCwd(chatId);
    switch (this.getChatProvider(chatId)) {
      case 'codex':
        return getCodexValidSessions(cwd, this.defaultCwd, limit);
      case 'claude':
      default:
        return getClaudeValidSessions(cwd, this.defaultCwd, limit);
    }
  }

  private getSessionListForChat(chatId: string): SessionTarget[] {
    const cwd = this.getChatCwd(chatId);
    switch (this.getChatProvider(chatId)) {
      case 'codex':
        return getCodexSessionList(cwd, this.defaultCwd);
      case 'claude':
      default:
        return getClaudeSessionList(cwd, this.defaultCwd);
    }
  }

  private formatSessionLine(index: number, session: SessionSummary, currentSession?: string): string {
    const marker = session.sessionId === currentSession ? ' 当前' : '';
    const age = formatDuration((Date.now() - session.mtimeMs) / 1000);
    return `**${index}.** \`${session.sessionId.slice(0, 8)}...\`${marker}  ${age}前`;
  }

  private sendResponse(chatId: string, text: string): void {
    const maxLen = 4000;
    if (text.length <= maxLen) {
      messageService.sendCardMessage(chatId, text).then(() => {
        logger.debug('[ChatManager] Response sent', { chatId, textLength: text.length });
      }).catch(err => {
        logger.error('[ChatManager] Failed to send response', { chatId, error: err.message });
      });
      return;
    }

    this.sendLongMessage(chatId, text, maxLen).catch(err => {
      logger.error('[ChatManager] Failed to send long response', { chatId, error: err.message });
    });
  }

  private async sendLongMessage(chatId: string, text: string, maxLen: number): Promise<void> {
    for (let i = 0; i < text.length; i += maxLen) {
      const chunk = text.slice(i, i + maxLen);
      try {
        await messageService.sendTextMessage(chatId, chunk);
      } catch (err: any) {
        logger.error('[ChatManager] Failed to send chunk', { chatId, error: err.message });
      }
    }
  }

  private async destroyAgent(chatId: string, reason: string): Promise<void> {
    const agent = this.agents.get(chatId);
    if (!agent) {
      return;
    }

    try {
      await agent.destroy();
    } catch (error: any) {
      logger.error(`[ChatManager] Error destroying agent while ${reason}`, {
        chatId,
        provider: this.getChatProvider(chatId),
        error: error.message,
      });
    }
    this.agents.delete(chatId);
  }
}

export const chatManager = new ChatManager();
